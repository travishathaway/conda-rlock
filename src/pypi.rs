use std::collections::HashMap;
use std::sync::Arc;

use indexmap::IndexMap;
use rattler_installs_packages::index::{
    ArtifactRequest, CheckAvailablePackages, PackageDb, PackageSourcesBuilder,
};
use rattler_installs_packages::normalize_index_url;
use rattler_installs_packages::python_env::Distribution;
use rattler_installs_packages::resolve::PypiVersion;
use rattler_installs_packages::types::{ArtifactInfo, ArtifactName, NormalizedPackageName};
use reqwest::Client;
use reqwest_middleware::ClientWithMiddleware;
use tokio::runtime::Runtime;
use url::Url;

/// Returns a new PackageDb instance for the PyPI we use in this project
pub fn get_package_db(index_url: &str) -> miette::Result<PackageDb> {
    let index_url = normalize_index_url(
        Url::parse(index_url)
            .map_err(|err| miette::miette!("failed to parse index URL: {:?}", err))?,
    );
    let sources = PackageSourcesBuilder::new(index_url).build()?;

    let cache_dir = dirs::cache_dir()
        .ok_or_else(|| miette::miette!("failed to determine cache directory"))?
        .join("rattler/pypi");

    let client = ClientWithMiddleware::from(Client::new());

    PackageDb::new(sources, client, &cache_dir, CheckAvailablePackages::Always)
}

/// Used to retrieve information from the index for the provided packages
pub fn get_available_artifacts(
    package_db: &PackageDb,
    packages: Vec<Distribution>,
) -> miette::Result<HashMap<NormalizedPackageName, Arc<ArtifactInfo>>> {
    let mut artifacts = HashMap::new();

    let runtime = Runtime::new().unwrap();

    runtime.block_on(async {
        for dist_info in packages {
            let request = ArtifactRequest::FromIndex(dist_info.name.clone());
            let available_artifacts = package_db.available_artifacts(request).await.unwrap();

            let artifact_name = dist_info.name.clone();
            if let Ok(matching_artifact) =
                match_distribution_with_artifact(dist_info.clone(), &available_artifacts)
            {
                artifacts.insert(artifact_name, matching_artifact);
            }
        }
    });

    Ok(artifacts)
}

/// A function that matches information provided as a `Distribution` struct with the data returned
/// from the `PackageDb::available_artifacts` method.
pub fn match_distribution_with_artifact(
    distribution: Distribution,
    artifacts: &IndexMap<PypiVersion, Vec<Arc<ArtifactInfo>>>,
) -> miette::Result<Arc<ArtifactInfo>> {
    let version = distribution.version.clone();
    let dist_version = PypiVersion::Version {
        version: version.clone(),
        package_allows_prerelease: version.any_prerelease(),
    };

    if let Some(matching_artifacts) = artifacts.get(&dist_version) {
        for matchng_art in matching_artifacts {
            let filename = matchng_art.filename.clone();
            if let ArtifactName::Wheel(_) = filename {
                return Ok(matchng_art.clone());
            }
        }
    }

    Err(miette::miette!(
        "no artifacts found for version: {:?}",
        distribution.version
    ))
    // if let Some(matching_artifacts) = artifacts.get(&dist_version) {
    //     matching_artifacts.iter().filter_map(|artifact| {
    //         if artifact.python_version.is_none() {
    //             Some(artifact.clone())
    //         } else {
    //             None
    //         }
    //     }).next().ok_or_else(|| miette::miette!("no artifacts found for version: {:?}", distribution.version))
    // } else {
    //     Err(miette::miette!("no artifacts found for version: {:?}", distribution.version))
    // }
}
