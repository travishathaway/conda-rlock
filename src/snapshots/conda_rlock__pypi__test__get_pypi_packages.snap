---
source: src/pypi.rs
expression: distributions
---
[
  Distribution(
    name: "keyring",
    version: "25.6.0",
    installer: Some("pip"),
    dist_info: "lib/python3.13/site-packages/keyring-25.6.0.dist-info",
    tags: Some([
      "py3-none-any",
    ]),
  ),
  Distribution(
    name: "werkzeug",
    version: "3.1.3",
    installer: Some("pip"),
    dist_info: "lib/python3.13/site-packages/werkzeug-3.1.3.dist-info",
    tags: Some([
      "py3-none-any",
    ]),
  ),
  Distribution(
    name: "itsdangerous",
    version: "2.2.0",
    installer: Some("pip"),
    dist_info: "lib/python3.13/site-packages/itsdangerous-2.2.0.dist-info",
    tags: Some([
      "py3-none-any",
    ]),
  ),
  Distribution(
    name: "blinker",
    version: "1.9.0",
    installer: Some("pip"),
    dist_info: "lib/python3.13/site-packages/blinker-1.9.0.dist-info",
    tags: Some([
      "py3-none-any",
    ]),
  ),
  Distribution(
    name: "flask",
    version: "3.1.0",
    installer: Some("pip"),
    dist_info: "lib/python3.13/site-packages/flask-3.1.0.dist-info",
    tags: Some([
      "py3-none-any",
    ]),
  ),
  Distribution(
    name: "jinja2",
    version: "3.1.5",
    installer: Some("pip"),
    dist_info: "lib/python3.13/site-packages/jinja2-3.1.5.dist-info",
    tags: Some([
      "py3-none-any",
    ]),
  ),
  Distribution(
    name: "jeepney",
    version: "0.8.0",
    installer: Some("pip"),
    dist_info: "lib/python3.13/site-packages/jeepney-0.8.0.dist-info",
    tags: Some([
      "py3-none-any",
    ]),
  ),
  Distribution(
    name: "markupsafe",
    version: "3.0.2",
    installer: Some("pip"),
    dist_info: "lib/python3.13/site-packages/MarkupSafe-3.0.2.dist-info",
    tags: Some([
      "cp313-cp313-manylinux_2_17_x86_64",
      "cp313-cp313-manylinux2014_x86_64",
    ]),
  ),
  Distribution(
    name: "cryptography",
    version: "44.0.1",
    installer: Some("pip"),
    dist_info: "lib/python3.13/site-packages/cryptography-44.0.1.dist-info",
    tags: Some([
      "cp39-abi3-manylinux_2_34_x86_64",
    ]),
  ),
  Distribution(
    name: "secretstorage",
    version: "3.3.3",
    installer: Some("pip"),
    dist_info: "lib/python3.13/site-packages/SecretStorage-3.3.3.dist-info",
    tags: Some([
      "py3-none-any",
    ]),
  ),
  Distribution(
    name: "jaraco-classes",
    version: "3.4.0",
    installer: Some("pip"),
    dist_info: "lib/python3.13/site-packages/jaraco.classes-3.4.0.dist-info",
    tags: Some([
      "py3-none-any",
    ]),
  ),
  Distribution(
    name: "click",
    version: "8.1.8",
    installer: Some("pip"),
    dist_info: "lib/python3.13/site-packages/click-8.1.8.dist-info",
    tags: Some([
      "py3-none-any",
    ]),
  ),
  Distribution(
    name: "pycparser",
    version: "2.22",
    installer: Some("pip"),
    dist_info: "lib/python3.13/site-packages/pycparser-2.22.dist-info",
    tags: Some([
      "py3-none-any",
    ]),
  ),
  Distribution(
    name: "jaraco-functools",
    version: "4.1.0",
    installer: Some("pip"),
    dist_info: "lib/python3.13/site-packages/jaraco.functools-4.1.0.dist-info",
    tags: Some([
      "py3-none-any",
    ]),
  ),
  Distribution(
    name: "more-itertools",
    version: "10.6.0",
    installer: Some("pip"),
    dist_info: "lib/python3.13/site-packages/more_itertools-10.6.0.dist-info",
    tags: Some([
      "py3-none-any",
    ]),
  ),
  Distribution(
    name: "jaraco-context",
    version: "6.0.1",
    installer: Some("pip"),
    dist_info: "lib/python3.13/site-packages/jaraco.context-6.0.1.dist-info",
    tags: Some([
      "py3-none-any",
    ]),
  ),
  Distribution(
    name: "cffi",
    version: "1.17.1",
    installer: Some("pip"),
    dist_info: "lib/python3.13/site-packages/cffi-1.17.1.dist-info",
    tags: Some([
      "cp313-cp313-manylinux_2_17_x86_64",
      "cp313-cp313-manylinux2014_x86_64",
    ]),
  ),
]
